File: 474-OnesandZeroes/474-OnesandZeroes.cpp
// Last updated: 11/11/2025, 2:11:17 pm
#include <bits/stdc++.h>
#include <iostream>
#include <vector>

using namespace std;

struct b_size{
    int ones = 0;
    int zeroes = 0;

    b_size() {

    }
};

class Solution {
public:
    int findMaxForm(vector<string>& strs, int m, int n) {
        int N = strs.size();

        vector<b_size> sizes(N);
        for (int i = 0; i < N; i++) {
            int one = 0;
            int zero = 0;

            for (int j = 0; j < strs[i].length(); j++) {
                if (strs[i][j] == '0') zero++;
                else one++;
            }

            sizes[i].ones = one;
            sizes[i].zeroes = zero;
        }
        int neg_inf = INT_MIN;

        int dp[N + 1][m + 1][n + 1];
        for (int i = 0; i <= N; i++) {
            for (int j = 0; j <= m; j++) {
                for (int k = 0; k <= n; k++) {
                    dp[i][j][k] = neg_inf;
                }
            }
        }
        dp[0][m][n] = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j <= m; j++) {
                for (int k = 0; k <= n; k++) {
                    int ones = sizes[i].ones;
                    int zeroes = sizes[i].zeroes;

                    dp[i + 1][j][k] = max(dp[i + 1][j][k], dp[i][j][k]);
                    if (j >= zeroes && k >= ones) {
                        dp[i + 1][j - zeroes][k - ones] = max(dp[i + 1][j - zeroes][k - ones], dp[i][j][k] + 1);
                    }
                }
            }
        }

        int ans = 0;
        for (int j = 0; j <= m; j++) {
            for (int k = 0; k <= n; k++) {
                ans = max(ans, dp[N][j][k]);
            }
        }

        return ans;
    }
};

File: 474-OnesandZeroes/474-OnesandZeroes.py
# Last updated: 11/11/2025, 1:57:23 pm
class Solution:
    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        N = len(strs)
        sizes = []
        
        for i in range(N):
            one = strs[i].count('1')
            zero = len(strs[i]) - one
            sizes.append((one, zero))
        
        @cache
        def rec(index, rem_m, rem_n):
            if index == N:
                return 0
            
            one = sizes[index][0]
            zero = sizes[index][1]

            ans = rec(index + 1, rem_m, rem_n)
            if rem_m >= zero and rem_n >= one:
                ans = max(ans, rec(index + 1, rem_m - zero, rem_n - one) + 1)
            
            return ans
        
        ans = rec(0, m, n)
        rec.cache_clear()
        return ans